ExampleCodes/LinearSolvers/GMRES/Poisson/AMReX_GMRES_Poisson.H

```diff
@@ -0,0 +1,193 @@
+    bool usePrecond (bool new_flag) { return std::exchange(m_use_precond, new_flag); }
+
+private:
+    GM m_gmres;
+    BoxArray m_ba;
+    DistributionMapping m_dm;
+    Geometry m_geom;
+    bool m_use_precond = false;
+};
+
+template <typename MF>
+GMRESPOISSONT<MF>::GMRESPOISSONT (const BoxArray& ba, const DistributionMapping& dm, const Geometry& geom)
+    : m_ba(ba), m_dm(dm), m_geom(geom)
+{
+    m_gmres.define(*this);
+}
+
+template <typename MF>
+auto GMRESPOISSONT<MF>::makeVecRHS () const -> MF
+{
+    return MultiFab(m_ba, m_dm, 1, 0);
+}
+
+template <typename MF>
+auto GMRESPOISSONT<MF>::makeVecLHS () const -> MF
+{
+    return MultiFab(m_ba, m_dm, 1, 1);
+}
+
+template <typename MF>
+auto GMRESPOISSONT<MF>::norm2 (MF const& mf) const -> RT
+{
+    return mf.norm2();
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::scale (MF& mf, RT scale_factor)
+{
+    mf.mult(scale_factor);
+}
+
+template <typename MF>
+auto GMRESPOISSONT<MF>::dotProduct (MF const& mf1, MF const& mf2) const -> RT
+{
+    return MultiFab::Dot(mf1,0,mf2,0,1,0);
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::setToZero (MF& lhs)
+{
+    lhs.setVal(0.);
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::assign (MF& lhs, MF const& rhs)
+{
+    MultiFab::Copy(lhs,rhs,0,0,1,0);
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::increment (MF& lhs, MF const& rhs, RT a)
+{
+    MultiFab::Saxpy(lhs,a,rhs,0,0,1,0);
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::linComb (MF& lhs, RT a, MF const& rhs_a, RT b, MF const& rhs_b)
+{
+    MultiFab::LinComb(lhs,a,rhs_a,0,b,rhs_b,0,0,1,0);
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::apply (MF& lhs, MF& rhs) const
+{
+    // apply matrix to rhs for output lhs
+    rhs.FillBoundary(m_geom.periodicity());
+
+    const GpuArray<Real, AMREX_SPACEDIM> dx = m_geom.CellSizeArray();
+
+    for ( MFIter mfi(rhs,TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+
+        const Box& bx = mfi.tilebox();
+
+        const Array4<const Real> & rhs_p = rhs.array(mfi);
+        const Array4<      Real> & lhs_p = lhs.array(mfi);
+
+        amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
+        {
+            lhs_p(i,j,k) = ( rhs_p(i+1,j,k) - 2.*rhs_p(i,j,k) + rhs_p(i-1,j,k) ) / (dx[0]*dx[0])
+                         + ( rhs_p(i,j+1,k) - 2.*rhs_p(i,j,k) + rhs_p(i,j-1,k) ) / (dx[1]*dx[1])
+#if (AMREX_SPACEDIM == 3)
+                         + ( rhs_p(i,j,k+1) - 2.*rhs_p(i,j,k) + rhs_p(i,j,k-1) ) / (dx[2]*dx[2])
+#endif
+                ;
+        });
+
+    }
+
+
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::precond (MF& lhs, MF const& rhs) const
+{
+    if (m_use_precond) {
+
+    } else {
+        MultiFab::Copy(lhs,rhs,0,0,1,0);
+    }
+}
+
+template <typename MF>
+void GMRESPOISSONT<MF>::solve (MF& a_sol, MF const& a_rhs, RT a_tol_rel, RT a_tol_abs)
+{
+    m_gmres.solve(a_sol, a_rhs, a_tol_rel, a_tol_abs);
+}
+
+using GMRESPOISSON = GMRESPOISSONT<MultiFab>;
```

Comment:
If we make this non-template, we can remove this.

Reply:
done